crt/fork.c

```diff
@@ -0,0 +1,569 @@
+        goto done;
+    }
+
+    if (!_within(child_stack, child_stack_size, cbp))
+    {
+        assert("child base pointer out of range" == NULL);
+        goto done;
+    }
+
+    for (size_t i = 0; pbp; i++)
+    {
+        *(uint64_t*)cbp -= delta;
+
+        pbp = *(void**)pbp;
+
+        if (!pbp)
+            break;
+
+        cbp = *(void**)cbp;
+
+        if (!_within(parent_stack, parent_stack_size, pbp))
+        {
+            // assert("current parent base pointer out of range" == NULL);
+            // goto done;
+            break;
+        }
+
+        if (!_within(child_stack, child_stack_size, cbp))
+        {
+            break;
+        }
+
+        assert((uint8_t*)cbp + delta == pbp);
+    }
+
+    *child_sp_out = (uint8_t*)parent_sp - delta;
+    *child_bp_out = (uint8_t*)parent_bp - delta;
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static int _child_func(void* arg)
+{
+    struct thread_args* args = (struct thread_args*)arg;
+    args->env.rsp = (uint64_t)args->child_sp;
+    args->env.rbp = (uint64_t)args->child_bp;
+
+    /* set the fsbase register to point to the child_td */
+    args->child_pthread->tid = getpid();
+    _set_fsbase(args->child_pthread);
+
+    /* set the pid that the parent is waiting on */
+    args->pid = getpid();
+
+    /* queue up the cleanup of these memory regions for this process exit */
+    /* We cannot safely free the new stack safely because the freeing will
+     * return to the same stack before next syscall to exit. */
+    syscall(
+        SYS_myst_munmap_on_exit,
+        args->unmap_on_exit[0].mmap_ptr,
+        args->unmap_on_exit[1].mmap_ptr_size);
+    syscall(
+        SYS_myst_munmap_on_exit,
+        args->unmap_on_exit[1].mmap_ptr,
+        args->unmap_on_exit[1].mmap_ptr_size);
+
+    /* jump back but on the new child stack */
+    myst_longjmp(&args->env, 1);
+    return 0;
+}
+
+__attribute__((__returns_twice__))
+__attribute__((__optimize__("-fno-stack-protector"))) pid_t
+myst_fork(void)
+{
+    pid_t pid = 0;
+    myst_jmp_buf_t env;
+
+    if (myst_setjmp(&env) == 0) /* parent */
+    {
+        struct thread_args* args;
+        size_t args_size;
+        const void* parent_sp = (const void*)env.rsp;
+        const void* parent_bp = (const void*)env.rbp;
+        void* sp = NULL;
+        void* bp = NULL;
+        const int clone_flags = CLONE_VM | CLONE_VFORK | SIGCHLD;
+        long tmp_ret;
+        struct pthread* child_pthread;
+        void* parent_stack;
+        void* stack;
+        size_t stack_size;
+        size_t parent_stack_size;
+
+        if (!(child_pthread = _create_child_pthread()))
+            return -ENOMEM;
+
+        stack = (uint8_t*)child_pthread->stack - child_pthread->stack_size;
+        stack_size = child_pthread->stack_size;
+
+        _get_parent_stack(&parent_stack, &parent_stack_size);
+
+        assert(stack_size == parent_stack_size);
+
+        if (_fixup_frame_pointers(
```

Comment:
Do we need to copy parent's stack?

Reply:
by the time we have got here we have created and copied the stack, and this is fixing up all the rbp pointers that allow the stack to be walked

Reply:
renamed